### PR TITLE
Add DeltaConfigs to support table properties stored in Metadata.configuration

### PR DESCRIPTION
**Pull request by allisonport-db**
This PR introduces `DeltaConfigs` which stores table configurations. Configurations have default values, and are able to be validated and created from a `Metadata.configuration`.

The PR introduces:
• `DeltaConfig.scala`
• `CalendarInterval.scala` (storing calendar time intervals)
• `IntervalUtils.scala` (parsing `CalendarInterval` formatted strings)
• `DateTimeConstants.java`
• `DeltaConfigSuite`

Changes were made in:
• `DeltaLogImpl.scala` (updated to use `DeltaConfigs`)
• `MetadataCleanup.scala` (updated to use `DeltaConfigs`)
• `OptimisticTransactionImpl.scala` (updated to use `DeltaConfigs.mergeGlobalConfigs`)
• `DeltaRetentionSuite.scala`
• `OptimisticTransactionSuite.scala`

### CODE REVIEW

File: standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.standalone.internal.util
+
+// todo: what privacy do we want?
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala:19]: `private[internal]` is fine - we don't have any APIs that expose this class publicly (and if we did, then this would have to be in Java).

also, this is fine in `internal.util` package. nice job :)

File: standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * The class representing calendar intervals. The calendar interval is stored internally in
+ * three components.
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala:22]: list the three components

File: standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * The `months` and `days` are not units of time with a constant length (unlike hours, seconds), so
+ * they are two separated fields from microseconds. One month may be equal to 28, 29, 30 or 31 days
+ * and one day may be equal to 23, 24 or 25 hours (daylight saving).
+ *
+ * @param months  an integer value representing the number of months in this interval
+ * @param days  an integer value representing the number of days in this interval
+ * @param microseconds  a long value representing the  number of microseconds in this interval
+ */
+private[internal] class CalendarInterval(val months: Int,
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala:32]: nit: indentation

File: standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                                         val days: Int,
+                                         val microseconds: Long) {
+
+  override def hashCode(): Int = {
+    // todo: implement more exactly? not used
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala:37]: yup, since we are overriding this method, we should implement it correctly. it's not used now but we never know if it will be used in the future.

File: standalone/src/main/scala/io/delta/standalone/internal/util/IntervalUtils.scala [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+ * limitations under the License.
+ */
+
+package io.delta.standalone.internal.util
+
+import java.nio.charset.StandardCharsets
+
+import io.delta.standalone.util.DateTimeConstants._
+
+object IntervalUtils {
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/util/IntervalUtils.scala:23]: nit: `private[internal]`

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala [hunk 6]
```diff
@@ -14,17 +14,256 @@
...
+import io.delta.standalone.util.DateTimeConstants
+import io.delta.standalone.internal.actions.{Action, Metadata, Protocol}
+import io.delta.standalone.internal.exception.DeltaErrors
+import io.delta.standalone.internal.util.{CalendarInterval, IntervalUtils}
 
-  // TODO CHECKPOINT_INTERVAL
+import org.apache.hadoop.conf.Configuration
 
-  // TODO: LOG_RETENTION
+case class DeltaConfig[T](
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala:28]: nit: private[internal]

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala [hunk 7]
```diff
@@ -14,17 +14,256 @@
...
+    validate(value)
+    key -> value
+  }
 }
+
+// TODO: do we want to extend and include DeltaLogging?
+/**
+ * Contains list of reservoir configs and validation checks.
+ */
+trait DeltaConfigsBase {
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala:71]: nit: private[internal]

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala [hunk 8]
```diff
@@ -14,17 +14,256 @@
...
+  val IS_APPEND_ONLY = buildConfig[Boolean](
+    "appendOnly",
+    "false",
+    _.toBoolean,
+    _ => true,
+    "needs to be a boolean.",
+    Some(new Protocol(0, 2)))
+}
+
+object DeltaConfigs extends DeltaConfigsBase
```
Comment (scottsand-db) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala:268]: nit: private[internal]

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  test("mergeGlobalConfigs") {
+
+    // todo: this is kind of tested in OptimisticTransactionSuite?
+    val hadoopConf = new Configuration()
+    hadoopConf.set("appendOnly", "false")
+    hadoopConf.set("enableExpiredLogCleanup", "true")
+    val metadataConf = Map("delta.enableExpiredLogCleanup"-> "false",
+      "delta.minWriterVersion" -> "0")
+    val mergedConf = DeltaConfigs.mergeGlobalConfigs(hadoopConf, metadataConf)
+    assert(mergedConf.contains("delta.appendOnly") &&
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala:40]: No need for `mergedConf.contains("delta.appendOnly")`. `mergedConf.get("delta.appendOnly")` is a valid call (will not throw exception) regardless of if the key exists or not.

If, for example, you did, `mergedConf("someKey")` then you would want to do a `.contains` check before, since map(key) calls do throw if the key doesn't exist.

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+import io.delta.standalone.internal.util.CalendarInterval
+
+// scalastyle:off funsuite
+import org.scalatest.FunSuite
+
+class DeltaConfigSuite extends FunSuite {
+
+  test("mergeGlobalConfigs") {
+
+    // todo: this is kind of tested in OptimisticTransactionSuite?
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala:33]: thanks for pointing this out. this is fine, keep both tests

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      mergedConf.get("delta.appendOnly") == Some("false"))
+    assert(mergedConf.contains("delta.enableExpiredLogCleanup") &&
+      mergedConf.get("delta.enableExpiredLogCleanup") == Some("false"))
+    assert(mergedConf.contains("delta.minWriterVersion") &&
+      mergedConf.get("delta.minWriterVersion") == Some("0"))
+  }
+
+  // todo: edge cases for parsing? add test cases?
+
+    test("parseCalendarInterval") {
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala:50]: nit: indent this back

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    hadoopConf.set("enableExpiredLogCleanup", "true")
+    val metadataConf = Map("delta.enableExpiredLogCleanup"-> "false",
+      "delta.minWriterVersion" -> "0")
+    val mergedConf = DeltaConfigs.mergeGlobalConfigs(hadoopConf, metadataConf)
+    assert(mergedConf.contains("delta.appendOnly") &&
+      mergedConf.get("delta.appendOnly") == Some("false"))
+    assert(mergedConf.contains("delta.enableExpiredLogCleanup") &&
+      mergedConf.get("delta.enableExpiredLogCleanup") == Some("false"))
+    assert(mergedConf.contains("delta.minWriterVersion") &&
+      mergedConf.get("delta.minWriterVersion") == Some("0"))
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala:45]: can you also check that non-set configs (e.g. tombstone retention) are set to their default values?

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  // todo: edge cases for parsing? add test cases?
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala:48]: sure. what happens if a user specifies a non-existent key, or passes in an invalid value?

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaRetentionSuite.scala [hunk 14]
```diff
@@ -135,12 +135,12 @@ class DeltaRetentionSuite extends DeltaRetentionSuiteBase {
     withTempDir { tempDir =>
       val log = DeltaLogImpl.forTable(hadoopConf, tempDir.getCanonicalPath)
       log.startTransaction().commit(
-          Metadata(configuration = Map("enableExpiredLogCleanup" -> "true")) :: Nil,
+          Metadata(configuration = Map("delta.enableExpiredLogCleanup" -> "true")) :: Nil,
```
[the remark this answers is not in the document]
Comment (scottsand-db) [reply standalone/src/test/scala/io/delta/standalone/internal/DeltaRetentionSuite.scala:138]: nice catch. yeah it is

**Review by scottsand-db — CHANGES_REQUESTED** [review]
Awesome work. Left mainly nits, answered the questions I saw, and added some comments in the tests.

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala (hunk 13 above)
Comment (allisonport-db) [reply standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala:44]: For non-existent keys: Isn't this not an issue until we call `DeltaConfigs.$CONFIG.fromMetaData(metadata)`? In which case we just want the default value (and to ignore the non-existent key)

For invalid values: I'm now realizing we don't validate the value (we don't invoke `apply` right?) But this is how OSS does it, when should we be validating the values?

Comment (scottsand-db) [reply standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala:44]: For non-specified keys, yes we just use the default value.

For an invalid key (e.g. `asdf`), I also suppose that is okay too. Since internally we only ever use DeltaConfigs.$CONFIG APIs, we will never use an invalid key.

Doing it the way OSS does it is how we should do it.

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala [hunk 15]
```diff
@@ -14,17 +14,231 @@
...
+    fromString: String => T,
+    validationFunction: T => Boolean,
+    helpMessage: String,
+    minimumProtocolVersion: Option[Protocol] = None,
+    editable: Boolean = true) {
+  /**
+   * Recover the saved value of this configuration from `Metadata`.
+   * If undefined, return defaultValue.
+   */
+  def fromMetaData(metadata: Metadata): T = {
```
Comment (allisonport-db) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala:38]: Why is this `fromMetaData` and not `fromMetadata`? Former is copied from OSS, but in connectors all references I've seen have been `Metadata`. Should this be changed to match the rest of the code?

File: standalone/src/main/java/io/delta/standalone/util/ParquetSchemaConverter.java [hunk 16]
```diff
@@ -0,0 +1,4 @@
+package io.delta.standalone.util;
+
+public class ParquetSchemaConverter {
```
Comment (zsxwing) [inline standalone/src/main/java/io/delta/standalone/util/ParquetSchemaConverter.java:3]: nit: remove this unused file?

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala [hunk 17]
```diff
@@ -14,17 +14,232 @@
...
+  def apply(value: String): (String, String) = {
+    validate(value)
+    key -> value
+  }
+}
+
+/**
+ * Contains list of reservoir configs and validation checks.
+ */
+private[internal] trait DeltaConfigsBase {
```
Comment (zsxwing) [inline standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala:69]: You can make this `object DeltaConfigs` directly.

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala [hunk 18]
```diff
@@ -14,17 +14,256 @@
...
+    entries.put(key.toLowerCase(Locale.ROOT), deltaConfig)
+    deltaConfig
+  }
+
+  /**
+   * Validates specified configurations and returns the normalized key -> value map.
+   */
+  def validateConfigurations(configurations: Map[String, String]): Map[String, String] = {
+    configurations.map {
+      case kv @ (key, value) if key.toLowerCase(Locale.ROOT).startsWith("delta.constraints.") =>
```
[the remark this answers is not in the document]
Comment (zsxwing) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala:119]: Right. We should throw an error.

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala [hunk 19]
```diff
@@ -14,17 +14,256 @@
...
+      case (key, value) if key.toLowerCase(Locale.ROOT).startsWith("delta.") =>
+        Option(entries.get(key.toLowerCase(Locale.ROOT).stripPrefix("delta.")))
+          .map(_(value))
+          .getOrElse {
+            throw DeltaErrors.unknownConfigurationKeyException(key)
+          }
+      case keyvalue @ (key, _) =>
+        if (entries.containsKey(key.toLowerCase(Locale.ROOT))) {
+          // TODO: how do we log in connectors?
+//        logConsole(
```
[the remark this answers is not in the document]
Comment (zsxwing) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala:131]: Yep. Logging is pretty helpful for debugging. We can use slf4j similar to https://github.com/delta-io/connectors/blob/master/hive/src/main/scala/io/delta/hive/DeltaInputFormat.scala#L59

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala [hunk 20]
```diff
@@ -14,17 +14,256 @@
...
+  /**
+   * Table properties for new tables can be specified through Hadoop configurations. This method
+   * checks to see if any of the configurations exist among the Hadoop configurations and merges
+   * them with the user provided configurations. User provided configs take precedence.
+   */
+  def mergeGlobalConfigs(hadoopConf: Configuration, tableConf: Map[String, String]):
+  Map[String, String] = {
+    import collection.JavaConverters._
+
+    // todo: is there a hadoop conf prefix?
```
[the remark this answers is not in the document]
Comment (zsxwing) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala:151]: In OSS Delta, when a user sets a SQL conf using the prefix `spark.databricks.delta.properties.defaults.` such as `spark.databricks.delta.properties.defaults.randomPrefixLength`, we will set the table property `delta.randomPrefixLength` to this value when creating a table.

We can use the same mechanism here: when a user sets a **Hadoop** conf using the prefix `spark.databricks.delta.properties.defaults.` such as `spark.databricks.delta.properties.defaults.randomPrefixLength`, we will set the table property `delta.randomPrefixLength` to this value when **creating a table**.

File: standalone/src/main/scala/io/delta/standalone/internal/exception/DeltaErrors.scala [hunk 21]
```diff
@@ -273,6 +273,15 @@ private[internal] object DeltaErrors {
       s"Parsed $nestType type:\n${nested.toPrettyJson}")
   }
 
+  def cannotModifyTableProperty(prop: String): Throwable =
+    throw new UnsupportedOperationException(
+      s"The Delta table configuration $prop cannot be specified by the user")
+
+  def unknownConfigurationKeyException(confKey: String): Throwable = {
+    // TODO: AnalysisException ?
```
Comment (zsxwing) [inline standalone/src/main/scala/io/delta/standalone/internal/exception/DeltaErrors.scala:281]: IllegalArgumentException looks good to me

File: standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+ * they are two separated fields from microseconds. One month may be equal to 28, 29, 30 or 31 days
+ * and one day may be equal to 23, 24 or 25 hours (daylight saving).
+ *
+ * @param months  an integer value representing the number of months in this interval
+ * @param days  an integer value representing the number of days in this interval
+ * @param microseconds  a long value representing the  number of microseconds in this interval
+ */
+private[internal] class CalendarInterval(val months: Int, val days: Int, val microseconds: Long) {
+
+  override def hashCode(): Int = {
```
Comment (zsxwing) [inline standalone/src/main/scala/io/delta/standalone/internal/util/CalendarInterval.scala:38]: Since we are in Scala, we can make CalendarInterval be a case class and remove `hashCode` and `equals`.

File: standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala (hunk 19 above)
[the remark this answers is not in the document]
Comment (allisonport-db) [reply standalone/src/main/scala/io/delta/standalone/internal/DeltaConfig.scala:141]: Should I add this now or wait for a future PR?

File: standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.standalone.internal
+
+import java.util.concurrent.TimeUnit
+
+import org.apache.hadoop.conf.Configuration
```
Comment (scottsand-db) [inline standalone/src/test/scala/io/delta/standalone/internal/DeltaConfigSuite.scala:21]: small nit: next sprint we will be adding checks or at least style-guide documentation for import order.
so, `org.apache....` imports should come after `io.delta.standalone` imports (separated by 1 line).

But this is fine for now, since I plan on fixing all import orders next sprint. Just an FYI for future work.
